Review python code snippets (#264)
alphagov/notifications-tech-docs: source/documentation/client_docs/_python.md

```python
response = notifications_client.send_sms_notification(

>     phone_number="+447900900123",

Reviewer: i think this (and all numbers in this file) should be TV numbers - with a double 7 at the beginning rather than 79

```suggestion
    phone_number="+447700900123",
```

Author: Good catch - I will replace them all at once 🙌🏼

Author: This is done now ☺️

The one-click unsubscribe URL will be added to the headers of your email. Email clients will use it to add an unsubscribe button.

```python

> one_click_unsubscribe_url = "https://pigeons.gov.uk/unsubscribe.html?opaque=123456789", # optional string, a URL

Reviewer 2: Can we consider keeping all the example domains as `example.com` just to help make it really obvious that it is not a real site, but an example that needs to be replaced?

For example:

```python
email_reply_to_id="ca4fdde7-2a67-4a6c-8393-62aa7245751f", # optional UUID string
```

You can leave out this argument if your service only has one reply-to email address, or you want to use the default email address.

#### Response

If the request to the client is successful, the client returns a `dict`:

```python
{
  "id": "201b576e-c09b-467b-9dfa-9c3b689ee730",  # required string - notification ID
  "reference": "your reference",  # optional string - reference you provided when sending the message
  "content": {
    "subject": "Your upcoming pigeon registration appointment",  # required string - message subject

>     "body": "Dear Amala\r\n\r\nYour pigeon registration appointment is scheduled for 2018-01-01 at 01:00PM.\r\n\r\nPlease bring:\r\n\n\n* passport\n* utility bill\n* other id\r\n\r\nYours,\r\nPigeon Affairs Bureau",  # required string - message content

Author: Yes, we do 😳

>     "body": "Dear Amala\r\n\r\nYour pigeon registration appointment is scheduled for 2018-01-01 at 01:00PM.\r\n\r\nPlease bring:\r\n\n\n* passport\n* utility bill\n* other id\r\n\r\nYours,\r\nPigeon Affairs Bureau",  # required string - message content

Author: I adapted the response snippets from real responses I got while using the Python client 🙌🏼

>     "body": "Dear Amala\r\n\r\nYour pigeon registration appointment is scheduled for 2018-01-01 at 01:00PM.\r\n\r\nPlease bring:\r\n\n\n* passport\n* utility bill\n* other id\r\n\r\nYours,\r\nPigeon Affairs Bureau",  # required string - message content

Reviewer: scary that it puts `\n` in some places (like between the bullet points) and `\r\n` in other places

>     "body": "Dear Amala\r\n\r\nYour pigeon registration appointment is scheduled for 2018-01-01 at 01:00PM.\r\n\r\nPlease bring:\r\n\n\n* passport\n* utility bill\n* other id\r\n\r\nYours,\r\nPigeon Affairs Bureau",  # required string - message content

Author: Yeah - is it because between bullet points it's not a new paragraph, and other places it is maybe? 🍋 🤔

```python
{
    "id": "740e5834-3a29-46b4-9a6f-16142fde533a",  # required string - notification ID

>     "reference": "your reference",  # optional string - reference you provided when sending the message

Author: They could be useful for schemas, no?

In schemas we expect some properties to always be a string, while others can be string or None.

Maybe we could change the wording from `optional string` to `string or None`?

>     "reference": "your reference",  # optional string - reference you provided when sending the message

Reviewer: hmm i had a play around with looking at a few options and think `optional string` might be fine for now (assuming the field is there, if the key may not exist at all i think we need to say so)

what do you mean by "useful for schemas"? validating responses from 3rd party APIs by schema isnt super typical (we don't do it with our 3rd parties, nor do we do it with our outbound responses outside of unit tests 🤔)

if we're thinking about working with schemas then that's the benefit of publishing an openapi spec, and that should be used directly

>     "reference": "your reference",  # optional string - reference you provided when sending the message

Reviewer: also worth thinking about how these kinds of things interact with, for example:

`"email_address": "amala@example.com",  # required string for emails`
`"one_click_unsubscribe": "https://example.com/unsubscribe.html?opaque=123456789", # optional string, email only - URL that you provided so your recipients can unsubscribe`

where they're _sometimes_ always present, and _sometimes_ sometimes present 🤯

    "email_address": "amala@example.com",  # required string for emails

>     "phone_number": "+447900900123",  # required string for text messages

Reviewer: oops, missed one
```suggestion
    "phone_number": "+447700900123",  # required string for text messages
```

- [`include_jobs`](#include-jobs-optional)

##### All messages - as an iterator

> This will return a Python iterator object. This object will yield one notification at a time. Once it has got through the first 250 notifications, it will request the next page from the API and start yielding from there, and so on, until it yields all your notifications.

Reviewer: is it useful to know how the iterator works under the hood?

```suggestion
This will return a Python iterator object which yields one notification at a time until it has yielded all your notifications.
```

Reviewer: i also thought about generator vs iterator, not sure which is more idiomatic

Author: Nice one, I will also change it in the other place we have that text 🙌🏼

Author: As to iterator vs. generator, not sure - whatever we go with though, we should be consistent ☺️

Author: > iterator is a more general concept: any object whose class has a __next__ method (next in Python 2) and an __iter__ method that does return self.

> Every generator is an iterator, but not vice versa. A generator is built by calling a function that has one or more yield expressions (yield statements, in Python 2.5 and earlier), and is an object that meets the previous paragraph's definition of an iterator.

> You may want to use a custom iterator, rather than a generator, when you need a class with somewhat complex state-maintaining behavior, or want to expose other methods besides __next__ (and __iter__ and __init__). Most often, a generator (sometimes, for sufficiently simple needs, a generator expression) is sufficient, and it's simpler to code because state maintenance (within reasonable limits) is basically "done for you" by the frame getting suspended and resumed.

https://stackoverflow.com/questions/2776829/difference-between-pythons-generators-and-iterators

So it looks our method is both an iterator, and a generator :D

Author: The change you suggested is done now 🙌🏼

```python
response = get_all_notifications_iterator()

# to get a list of all notifications from your iterator:

> notifications = [notification for notification in response]

Author: Ooh, I like that, I can change it 🙌🏼

> notifications = [notification for notification in response]

Author: This is done now 🙌🏼

```python
response = notifications_client.post_template_preview(
    template_id="f33517ff-2a88-4f6e-b855-c550268ce08a",
    personalisation={

>     "first_name": "Amala",
>     "appointment_date": "2018-01-01 at 01:00PM",
>   },

Reviewer: minor formatting suggestions (indentation not consistent)

```suggestion
        "first_name": "Amala",
        "appointment_date": "2018-01-01 at 01:00PM",
    }
```

  "received_text_messages":
  [
    {
      "id": "'b51f638b-4295-46e0-a06e-cd41eee7c33b", # required string - ID of received text message
      "user_number": "447900900123", # required string - number of the end user who sent the message
      "notify_number": "07900900456", # required string - your receiving number

>       "created_at": "2024-12-12 18:39:16.123346"", # required string - date and time template created

Reviewer: typo
```suggestion
      "created_at": "2024-12-12 18:39:16.123346", # required string - date and time template created
```
